python/ray/train/v2/_internal/state/state_actor.py

```diff
@@ -1,47 +1,164 @@
     EventLogType,
     check_export_api_enabled,
     get_export_event_logger,
 )
 from ray.actor import ActorHandle
-from ray.train.v2._internal.state.schema import TrainRun, TrainRunAttempt
+from ray.train.v2._internal.constants import (
+    CONTROLLERS_TO_RECONCILE_PER_CHECK,
+    DEFAULT_ENABLE_STATE_ACTOR_POLLING,
+    DEFAULT_STATE_ACTOR_POLL_INTERVAL_S,
+    ENABLE_STATE_ACTOR_POLLING_ENV_VAR,
+    GET_ACTOR_TIMEOUT_S,
+    STATE_ACTOR_POLL_INTERVAL_S_ENV_VAR,
+)
+from ray.train.v2._internal.state.schema import (
+    TrainRun,
+    TrainRunAttempt,
+)
+from ray.train.v2._internal.state.util import (
+    is_actor_alive,
+    update_train_run_aborted,
+    update_train_run_attempt_aborted,
+)
+from ray.train.v2._internal.util import time_monotonic
 
 logger = logging.getLogger(__name__)
 
 
 class TrainStateActor:
-    def __init__(self):
+    def __init__(
+        self,
+        enable_state_actor_polling: bool = False,
+        poll_interval_s: float = 30,
+        get_actor_timeout_s: int = GET_ACTOR_TIMEOUT_S,
+        controllers_to_poll_per_iteration: int = CONTROLLERS_TO_RECONCILE_PER_CHECK,
+    ):
         # NOTE: All runs and attempts are stored in memory.
         # This may be a memory issue for large runs.
-        self._runs: Dict[str, TrainRun] = {}
+        # TODO: consider cleaning up runs over time.
+        self._runs: Dict[str, TrainRun] = OrderedDict()
         # {run_id: {attempt_id: TrainRunAttempt}}
-        self._run_attempts: Dict[str, Dict[str, TrainRunAttempt]] = defaultdict(dict)
+        self._run_attempts: Dict[str, OrderedDict[str, TrainRunAttempt]] = defaultdict(
+            OrderedDict
+        )
         (
             self._export_logger,
             self._is_train_run_export_api_enabled,
             self._is_train_run_attempt_export_api_enabled,
         ) = self._init_export_logger()
 
+        # TODO: consider row level locking if loop takes too long.
+        self._runs_lock = threading.RLock()
+        self._run_attempts_lock = threading.RLock()
+
+        # Set env vars related to polling the controller.
+        if enable_state_actor_polling:
+            self._poll_interval_s = poll_interval_s
+            self._controllers_to_poll_per_iteration = controllers_to_poll_per_iteration
+            self._get_actor_timeout_s = get_actor_timeout_s
+            self._start_run_state_reconciliation_thread()
+
+    def _abort_live_runs_with_dead_controllers(
+        self, last_poll_run_id: Optional[str]
+    ) -> str:
+        aborted_run_ids = []
+        with self._runs_lock:
+            runs = list(self._runs.values())
+
+            # Start iterating from poll index.
+            poll_index = 0
+            if last_poll_run_id is not None:
+                for i, run in enumerate(runs):
+                    if run.id == last_poll_run_id:
+                        poll_index = (i + 1) % len(runs)
+                        break
+
+            # Abort runs.
+            num_sampled_runs = 0
```

Comment:
Name this to `num_polled_runs`? For consistency unless there's a difference between sample and poll.
```suggestion
            num_polled_runs = 0
```

Reply:
Done - agreed the wording is confusing. The new wording is:
* Sample: removed. It was there initially because I was sampling runs randomly, which I'm not doing anymore.
* Reconcile: function names and variables that are "functional" now have the word "reconcile" in them because their goal is to reconcile the states of the train runs e.g. ENABLE_STATE_ACTOR_RECONCILIATION_ENV_VAR, STATE_ACTOR_RECONCILIATION_INTERVAL_S_ENV_VAR, _start_run_state_reconciliation_thread
* Poll: now refers to everything that checks individual runs e.g. poll_index, num_polled_runs, latest_poll_time
Lmk if this makes sense.